text/0058-valuecastable-format.md

```diff
@@ -0,0 +1,93 @@
+- Start Date: 2024-03-18
+- RFC PR: [amaranth-lang/rfcs#58](https://github.com/amaranth-lang/rfcs/pull/58)
+- Amaranth Issue: [amaranth-lang/amaranth#0000](https://github.com/amaranth-lang/amaranth/issues/0000)
+
+# Core support for `ValueCastable` formatting
+
+## Summary
+[summary]: #summary
+
+`Format` hooks are added, allowing custom formatting to be implemented for `ValueCastable`s.
+
+This RFC is only about adding hook support for custom `ShapeCastable`s. Providing actual formatting implementation for `lib.data` and `lib.enum` is left for a future RFC.
+
+## Motivation
+[motivation]: #motivation
+
+Custom types, like enums and data structures, will not make immediate sense to the user when printed as raw bit patterns. However, it is often desirable to print them out with the `Print` statement. Thus, an extension point for custom formatting needs to be added to the `Format` machinery.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`ShapeCastable` subtypes can define their own ways of formatting values. For example, consider the following:
+
+```py
+class Opcode(amaranth.lib.enum.Enum, shape=unsigned(2)):
+    ADD = 0
+    SUB = 1
+    XOR = 2
+
+op = Signal(Opcode)
+m.d.comb += [
+    op.eq(Opcode.XOR)
+    Print(Format("current opcode: {}", op)),
+]
+```
+
+It is desirable that this example would print something like `"XOR"`, instead of `"2"`. Such functionality can be added on `ShapeCastable` subclasses by overriding the `format` method.
+
+However, sometimes it is also useful to print the raw value of such a signal. A shorthand is provided for that purpose:
+
+```py
+m.d.comb += Print(Format("current opcode: {op} (raw: {op!v})", op=op))
```

Comment:
Bikeshed: I feel weird about `!v`. I guess it's `v` for `Value.cast`, but I feel that `!i` for "turn it into an Integer" or `!n` for "turn it into a Number` might look less out of place.

Reply:
`!r` would be perfect.

Reply:
... `!R`?

Reply:
or `!av` for `.as_value()`

Reply:
or `!l` for "lower to a value"

Reply:
`.as_value()` is distinct from `Value.cast` (it can return another `ValueCastable`)

... actually this may be legitimate alternate semantics